// @agct/react-common/dist/components/Accordion/Accordion.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', './Accordion'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('./Accordion'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.Accordion);
    global.AccordionTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _Accordion) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _Accordion2 = _interopRequireDefault(_Accordion);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());

  describe('Accordion component', function () {
    var props = {
      accordionHeaderTemplate: _react2.default.createElement(
        'div',
        { className: 'header-template' },
        'sds'
      ),
      accordionBodyTemplate: _react2.default.createElement(
        'div',
        { className: 'body-template' },
        'sss'
      ),
      className: 'custom-class',
      isOpen: false
    };

    it('renders header template', function () {
      var wrapper = (0, _enzyme.shallow)(_react2.default.createElement(_Accordion2.default, props));
      (0, _chai.expect)(wrapper.find('.header-template')).to.exist;
    });

    it('sets custom class prop', function () {
      var wrapper = (0, _enzyme.shallow)(_react2.default.createElement(_Accordion2.default, props));
      (0, _chai.expect)(wrapper.find('.custom-class')).to.exist;
    });
  });
});
