// @agct/reacceptance/karma.conf.js
module.exports = (config) => {
  config.set({
    basePath: '',
    browsers: ['PhantomJS'],
    frameworks: ['mocha', 'sinon'],
    reporters: ['spec', 'coverage', 'tap'],
    singleRun: true,
    port: 9879,

    files: [
      'node_modules/babel-polyfill/dist/polyfill.js',
      'src/**/*.test.+(js|jsx)',
    ],

    preprocessors: {
      'src/**/*.test.+(js|jsx)': ['webpack', 'sourcemap'],
    },

    webpack: {
      devtool: '#inline-source-map',
      module: {
        loaders: [
          {
            test: /\.jsx?$/,
            exclude: /node_modules\/(?!@agct)/,
            loader: 'babel',
          },
          {
            test: /\.json$/,
            loader: 'json',
          },
        ],
        postLoaders: [
          {
            test: /\.jsx?$/,
            exclude: /(\.test\.jsx?$|node_modules\/|tests\/)/,
            loader: 'istanbul-instrumenter',
          },
        ],
      },
      resolve: {
        modulesDirectories: [
          'src',
          'node_modules',
        ],
        extensions: ['', '.json', '.js', '.jsx'],
      },
      externals: {
        'react/addons': true,
        'react/lib/ExecutionEnvironment': true,
        'react/lib/ReactContext': true,
      },
    },

    webpackServer: {
      noInfo: true,
    },

    coverageReporter: {
      reporters: [
        { type: 'html', subdir: 'report-html' },
        { type: 'text-summary', subdir: '.', file: 'text-summary.txt' },
      ],
      check: {
        each: {
          statements: 0,
          functions: 0,
          branches: 0,
          lines: 0,
        },
      },
    },

    tapReporter: {
      outputFile: './results.tap',
      disableStdout: true,
    },

    plugins: [
      'karma-mocha',
      'karma-sinon',
      'karma-spec-reporter',
      'karma-coverage',
      'karma-tap-reporter',
      'karma-webpack',
      'karma-sourcemap-loader',
      'karma-chrome-launcher',
      'karma-phantomjs-launcher',
    ],
  });
};
