// @agct/react-common/src/components/ProgressIndicators/ProgressIndicators.jsx
import React, { PropTypes } from 'react';
import classNames from 'classnames';

export const SIZE_SMALL = 'small';
export const SIZE_LARGE = 'large';

export const ProgressIndicatorsItem = ({ title, description, isComplete, isActive }) => {
  const baseClass = classNames('progress__step', {
    'is-complete': isComplete,
    'is-active': isActive,
  });

  return (
    <div className="progress__item">
      <div className={baseClass}>
        <div className="progress__title">
          {title}
          {isComplete && <i className="agi agi-tick"></i>}
        </div>
        <div className="progress__description">{description}</div>
      </div>

      <div className="progress__indicator">
        <i className="agi agi-arrow-right"></i>
      </div>
    </div>
  );
};

ProgressIndicatorsItem.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  isComplete: PropTypes.bool,
  isActive: PropTypes.bool,
};

const ProgressIndicators = ({ size, children }) => {
  const baseClass = classNames('progress', {
    'progress--small': size === SIZE_SMALL,
    'progress--large': size === SIZE_LARGE,
  });
  return (
    <div className={baseClass}>
      {children}
    </div>
  );
};

ProgressIndicators.propTypes = {
  children: PropTypes.any,
  size: PropTypes.string,
};

export default ProgressIndicators;
