// @agct/react-common/dist/components/FeedbackLink/FeedbackLink.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react);
    global.FeedbackLink = mod.exports;
  }
})(this, function (exports, _react) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var generateEmail = function generateEmail(userAgent, username, date, url) {
    return '%0A-------------------------------------------------------%0AYour feedback is greatly appreciated.%0AIf you are experiencing a problem or have a suggestion, please be as specific as possible and include a screenshot where necessary.%0A%0ATechnical Data:%0Auser-agent: ' + userAgent + ' url: ' + url + ' screen-resolution: ' + window.outerWidth + 'x' + window.outerHeight + ' username: ' + username + ' timestamp: ' + date;
  };

  var sendEmail = function sendEmail(app, userAgent, username, date, url) {
    window.open('mailto:icebreakersfeedback@allangray.co.za?Subject=[' + app + ' Feedback]\n    &body=' + generateEmail(userAgent, username, date, url), 'emailWindow', '_self');
  };

  var FeedbackLink = function FeedbackLink(_ref) {
    var app = _ref.app;
    var userAgent = _ref.userAgent;
    var username = _ref.username;
    var _ref$date = _ref.date;
    var date = _ref$date === undefined ? new Date() : _ref$date;
    var _ref$url = _ref.url;
    var url = _ref$url === undefined ? window.location.href : _ref$url;

    var onClick = function onClick(e) {
      e.preventDefault();
      sendEmail(app, userAgent, username, date, url);
    };

    return _react2.default.createElement(
      'a',
      {
        className: 'nav-bar__items__item',
        onClick: onClick
      },
      'Send us your Feedback'
    );
  };

  FeedbackLink.propTypes = {
    app: _react.PropTypes.string.isRequired,
    userAgent: _react.PropTypes.string.isRequired,
    username: _react.PropTypes.string.isRequired,
    date: _react.PropTypes.object,
    url: _react.PropTypes.string
  };

  exports.default = FeedbackLink;
});
