// @agct/react-common/dist/components/ProgressIndicators/ProgressIndicators.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', 'components/ProgressIndicators/ProgressIndicators'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('components/ProgressIndicators/ProgressIndicators'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.ProgressIndicators);
    global.ProgressIndicatorsTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _ProgressIndicators) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _ProgressIndicators2 = _interopRequireDefault(_ProgressIndicators);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());

  describe('ProgressIndicators component', function () {
    var props = {
      title: 'step 1',
      description: 'step description'
    };

    it('renders small progressIndicators', function () {
      (0, _chai.expect)((0, _enzyme.shallow)(_react2.default.createElement(_ProgressIndicators2.default, { size: _ProgressIndicators.SIZE_SMALL })).find('.progress--small')).to.exist;
    });

    it('renders large progressIndicators', function () {
      (0, _chai.expect)((0, _enzyme.shallow)(_react2.default.createElement(_ProgressIndicators2.default, { size: _ProgressIndicators.SIZE_LARGE })).find('.progress--large')).to.exist;
    });

    it('renders children', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_ProgressIndicators2.default, {
        children: _react2.default.createElement(_ProgressIndicators.ProgressIndicatorsItem, { title: props.title, description: props.description })
      }));
      (0, _chai.expect)(wrapper.find('.progress__item')).to.exist;
      (0, _chai.expect)(wrapper.find('.progress__title').text()).to.equal(props.title);
      (0, _chai.expect)(wrapper.find('.progress__description').text()).to.equal(props.description);
    });

    it('renders completed item', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_ProgressIndicators2.default, {
        children: _react2.default.createElement(_ProgressIndicators.ProgressIndicatorsItem, { title: props.title, description: props.description, isComplete: true })
      }));

      (0, _chai.expect)(wrapper.find('.progress__step.is-complete')).to.exist;
      (0, _chai.expect)(wrapper.find('.agi-tick')).to.exist;
    });

    it('renders active item', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_ProgressIndicators2.default, {
        children: _react2.default.createElement(_ProgressIndicators.ProgressIndicatorsItem, { title: props.title, description: props.description, isActive: true })
      }));

      (0, _chai.expect)(wrapper.find('.progress__step.is-active')).to.exist;
    });
  });
});
