// @agct/react-common/dist/components/ProgressIndicators/ProgressIndicators.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'classnames'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames);
    global.ProgressIndicators = mod.exports;
  }
})(this, function (exports, _react, _classnames) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.ProgressIndicatorsItem = exports.SIZE_LARGE = exports.SIZE_SMALL = undefined;

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var SIZE_SMALL = exports.SIZE_SMALL = 'small';
  var SIZE_LARGE = exports.SIZE_LARGE = 'large';

  var ProgressIndicatorsItem = exports.ProgressIndicatorsItem = function ProgressIndicatorsItem(_ref) {
    var title = _ref.title;
    var description = _ref.description;
    var isComplete = _ref.isComplete;
    var isActive = _ref.isActive;

    var baseClass = (0, _classnames2.default)('progress__step', {
      'is-complete': isComplete,
      'is-active': isActive
    });

    return _react2.default.createElement(
      'div',
      { className: 'progress__item' },
      _react2.default.createElement(
        'div',
        { className: baseClass },
        _react2.default.createElement(
          'div',
          { className: 'progress__title' },
          title,
          isComplete && _react2.default.createElement('i', { className: 'agi agi-tick' })
        ),
        _react2.default.createElement(
          'div',
          { className: 'progress__description' },
          description
        )
      ),
      _react2.default.createElement(
        'div',
        { className: 'progress__indicator' },
        _react2.default.createElement('i', { className: 'agi agi-arrow-right' })
      )
    );
  };

  ProgressIndicatorsItem.propTypes = {
    title: _react.PropTypes.string.isRequired,
    description: _react.PropTypes.string.isRequired,
    isComplete: _react.PropTypes.bool,
    isActive: _react.PropTypes.bool
  };

  var ProgressIndicators = function ProgressIndicators(_ref2) {
    var size = _ref2.size;
    var children = _ref2.children;

    var baseClass = (0, _classnames2.default)('progress', {
      'progress--small': size === SIZE_SMALL,
      'progress--large': size === SIZE_LARGE
    });
    return _react2.default.createElement(
      'div',
      { className: baseClass },
      children
    );
  };

  ProgressIndicators.propTypes = {
    children: _react.PropTypes.any,
    size: _react.PropTypes.string
  };

  exports.default = ProgressIndicators;
});
