// @agct/react-common/src/components/Datepicker/DatepickerUtils.js
export const getTitleDayView = (moment) => (month, year) => moment().month(month).year(year).format('MMMM YYYY');

export const getTitleMonthView = (moment) => (year) => moment().year(year).format('YYYY');

export const getTitleYearView = (moment) => (startYear, endYear) => (
  `${moment().year(startYear).format('YYYY')}-${moment().year(endYear).format('YYYY')}`
);

export const getYearRange = (moment) => (year) => {
  const selectedYear = moment().year(year);
  const start = selectedYear.clone().subtract(7, 'years');
  const end = selectedYear.clone().add(4, 'years');
  return { start, end };
};

export const getDays = (moment) => (month, year) => {
  const selectedMonth = moment().year(year).month(month);
  const start = selectedMonth.clone().date(1).startOf('week');
  const end = selectedMonth.clone().endOf('month').endOf('week');
  return moment.range(start, end).toArray('days');
};

export const getMonths = (moment) => (year) => {
  const selectedYear = moment().year(year);
  const start = selectedYear.clone().startOf('year');
  const end = selectedYear.clone().endOf('year');
  return moment.range(start, end).toArray('months');
};

export const getYears = (moment) => (year) => {
  const getYearRangeMoment = getYearRange(moment);
  const { start, end } = getYearRangeMoment(year);
  return moment.range(start, end).toArray('years');
};
