// @agct/react-common/dist/utils/validators/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './isNone', './isBlank', './isEmail', './isEmpty', './isEqual', './isPresent', './isTooShort', './isTooLong'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./isNone'), require('./isBlank'), require('./isEmail'), require('./isEmpty'), require('./isEqual'), require('./isPresent'), require('./isTooShort'), require('./isTooLong'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.isNone, global.isBlank, global.isEmail, global.isEmpty, global.isEqual, global.isPresent, global.isTooShort, global.isTooLong);
    global.index = mod.exports;
  }
})(this, function (exports, _isNone, _isBlank, _isEmail, _isEmpty, _isEqual, _isPresent, _isTooShort, _isTooLong) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.isTooShort = exports.isTooLong = exports.isPresent = exports.isEqual = exports.isEmpty = exports.isEmail = exports.isBlank = exports.isNone = undefined;

  var _isNone2 = _interopRequireDefault(_isNone);

  var _isBlank2 = _interopRequireDefault(_isBlank);

  var _isEmail2 = _interopRequireDefault(_isEmail);

  var _isEmpty2 = _interopRequireDefault(_isEmpty);

  var _isEqual2 = _interopRequireDefault(_isEqual);

  var _isPresent2 = _interopRequireDefault(_isPresent);

  var _isTooShort2 = _interopRequireDefault(_isTooShort);

  var _isTooLong2 = _interopRequireDefault(_isTooLong);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.isNone = _isNone2.default;
  exports.isBlank = _isBlank2.default;
  exports.isEmail = _isEmail2.default;
  exports.isEmpty = _isEmpty2.default;
  exports.isEqual = _isEqual2.default;
  exports.isPresent = _isPresent2.default;
  exports.isTooLong = _isTooLong2.default;
  exports.isTooShort = _isTooShort2.default;
});
