// @agct/react-common/dist/components/Datepicker/DatepickerUtils.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.DatepickerUtils = mod.exports;
  }
})(this, function (exports) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  var getTitleDayView = exports.getTitleDayView = function getTitleDayView(moment) {
    return function (month, year) {
      return moment().month(month).year(year).format('MMMM YYYY');
    };
  };

  var getTitleMonthView = exports.getTitleMonthView = function getTitleMonthView(moment) {
    return function (year) {
      return moment().year(year).format('YYYY');
    };
  };

  var getTitleYearView = exports.getTitleYearView = function getTitleYearView(moment) {
    return function (startYear, endYear) {
      return moment().year(startYear).format('YYYY') + '-' + moment().year(endYear).format('YYYY');
    };
  };

  var getYearRange = exports.getYearRange = function getYearRange(moment) {
    return function (year) {
      var selectedYear = moment().year(year);
      var start = selectedYear.clone().subtract(7, 'years');
      var end = selectedYear.clone().add(4, 'years');
      return { start: start, end: end };
    };
  };

  var getDays = exports.getDays = function getDays(moment) {
    return function (month, year) {
      var selectedMonth = moment().year(year).month(month);
      var start = selectedMonth.clone().date(1).startOf('week');
      var end = selectedMonth.clone().endOf('month').endOf('week');
      return moment.range(start, end).toArray('days');
    };
  };

  var getMonths = exports.getMonths = function getMonths(moment) {
    return function (year) {
      var selectedYear = moment().year(year);
      var start = selectedYear.clone().startOf('year');
      var end = selectedYear.clone().endOf('year');
      return moment.range(start, end).toArray('months');
    };
  };

  var getYears = exports.getYears = function getYears(moment) {
    return function (year) {
      var getYearRangeMoment = getYearRange(moment);

      var _getYearRangeMoment = getYearRangeMoment(year);

      var start = _getYearRangeMoment.start;
      var end = _getYearRangeMoment.end;

      return moment.range(start, end).toArray('years');
    };
  };
});
