// @agct/react-common/dist/components/Breadcrumbs/Breadcrumbs.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'react-router', '@agct/reacceptance/src/qa'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-router'), require('@agct/reacceptance/src/qa'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactRouter, global.qa);
    global.Breadcrumbs = mod.exports;
  }
})(this, function (exports, _react, _reactRouter, _qa) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.BreadcrumbsItem = exports.Breadcrumbs = undefined;

  var _react2 = _interopRequireDefault(_react);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var Breadcrumbs = exports.Breadcrumbs = function Breadcrumbs(_ref) {
    var children = _ref.children;
    return _react2.default.createElement(
      'ul',
      _extends({ className: 'breadcrumbs' }, (0, _qa.attr)('breadcrumb-list')),
      children
    );
  };

  Breadcrumbs.propTypes = {
    children: _react.PropTypes.any
  };

  var BreadcrumbsItem = exports.BreadcrumbsItem = function BreadcrumbsItem(_ref2) {
    var title = _ref2.title;
    var url = _ref2.url;
    return _react2.default.createElement(
      'li',
      { className: 'breadcrumbs__breadcrumb' },
      url && _react2.default.createElement(
        'span',
        null,
        _react2.default.createElement(
          _reactRouter.Link,
          _extends({ to: url, className: 'breadcrumbs__breadcrumb__link' }, (0, _qa.attr)('breadcrumb-item')),
          title
        )
      ),
      !url && title
    );
  };

  BreadcrumbsItem.propTypes = {
    title: _react.PropTypes.string.isRequired,
    url: _react.PropTypes.string
  };

  exports.default = Breadcrumbs;
});
