// @agct/react-common/dist/utils/formatting/money.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['chai', './money'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('chai'), require('./money'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.chai, global.money);
    global.moneyTest = mod.exports;
  }
})(this, function (_chai, _money) {
  'use strict';

  describe('fixCurrencyCode', function () {
    it('converts YEN to JPY', function () {
      (0, _chai.expect)((0, _money.fixCurrencyCode)('YEN')).to.be.equal('JPY');
    });

    it('leaves JPY and other currency codes unaffected', function () {
      (0, _chai.expect)((0, _money.fixCurrencyCode)('JPY')).to.be.equal('JPY');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('USD')).to.be.equal('USD');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('GBP')).to.be.equal('GBP');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('ZAR')).to.be.equal('ZAR');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('AUD')).to.be.equal('AUD');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('AUD')).to.be.equal('AUD');
      (0, _chai.expect)((0, _money.fixCurrencyCode)('EUR')).to.be.equal('EUR');

      (0, _chai.expect)((0, _money.fixCurrencyCode)('XXX')).to.be.equal('XXX');
    });
  });

  describe('formatMoney', function () {
    it('default symbol is R', function () {
      (0, _chai.expect)((0, _money.formatMoney)(100.00)).to.be.equal('R 100.00');
    });

    it('thousand seperator and decimal characters', function () {
      (0, _chai.expect)((0, _money.formatMoney)(85432342341.98)).to.be.equal('R 85 432 342 341.98');
      (0, _chai.expect)((0, _money.formatMoney)(200.45)).to.be.equal('R 200.45');
      (0, _chai.expect)((0, _money.formatMoney)(74389.10)).to.be.equal('R 74 389.10');
    });

    it('negative number', function () {
      (0, _chai.expect)((0, _money.formatMoney)(-748302128.32)).to.be.equal('R-748 302 128.32');
    });

    it('zero number', function () {
      (0, _chai.expect)((0, _money.formatMoney)(0)).to.be.equal('R 0.00');
    });

    it('two decimal points', function () {
      (0, _chai.expect)((0, _money.formatMoney)(1.1)).to.be.equal('R 1.10');
      (0, _chai.expect)((0, _money.formatMoney)(57984398)).to.be.equal('R 57 984 398.00');
      (0, _chai.expect)((0, _money.formatMoney)(489.1234567)).to.be.equal('R 489.12');
    });

    it('does not round .004999', function () {
      (0, _chai.expect)((0, _money.formatMoney)(99.004999)).to.be.equal('R 99.00');
    });

    it('rounds .005 up', function () {
      (0, _chai.expect)((0, _money.formatMoney)(99.005)).to.be.equal('R 99.01');
    });

    it('rounds .995 up to wholde number', function () {
      (0, _chai.expect)((0, _money.formatMoney)(99.995)).to.be.equal('R 100.00');
    });
  });

  describe('formatMoneyWithSymbol', function () {
    it('ZAR', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(99, 'ZAR')).to.be.equal('R 99.00');
    });

    it('USD', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(74239.34, 'USD')).to.be.equal('$ 74 239.34');
    });

    it('AUD', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(9443471.48, 'AUD')).to.be.equal('A$ 9 443 471.48');
    });

    it('GBP', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(1, 'GBP')).to.be.equal('£ 1.00');
    });

    it('EUR', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(2890.32, 'EUR')).to.be.equal('€ 2 890.32');
    });

    it('JPY', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(99, 'JPY')).to.be.equal('¥ 99.00');
    });

    it('FLEX legacy YEN codes', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(99, 'YEN')).to.be.equal('¥ 99.00');
    });

    it('unknown XXX codes falls back', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(99, 'XXX')).to.be.equal('XXX 99.00');
    });

    it('YEN and JPY are formated with decimals due to FLEX legacy complications', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(983472.37123, 'YEN')).to.be.equal('¥ 983 472.37');
      (0, _chai.expect)((0, _money.formatMoneyWithSymbol)(90734578, 'JPY')).to.be.equal('¥ 90 734 578.00');
    });
  });

  describe('formatMoneyWithCode', function () {
    it('ZAR', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(99, 'ZAR')).to.be.equal('ZAR 99.00');
    });

    it('USD', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(74239.34, 'USD')).to.be.equal('USD 74 239.34');
    });

    it('AUD', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(9443471.48, 'AUD')).to.be.equal('AUD 9 443 471.48');
    });

    it('GBP', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(1, 'GBP')).to.be.equal('GBP 1.00');
    });

    it('EUR', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(2890.32, 'EUR')).to.be.equal('EUR 2 890.32');
    });

    it('JPY', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(99, 'JPY')).to.be.equal('JPY 99.00');
    });

    it('FLEX legacy YEN code converts to JPY', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(99, 'YEN')).to.be.equal('JPY 99.00');
    });

    it('unknown XXX codes falls back', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(99, 'XXX')).to.be.equal('XXX 99.00');
    });

    it('YEN and JPY are formated with decimals due to FLEX legacy complications', function () {
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(983472.37123, 'YEN')).to.be.equal('JPY 983 472.37');
      (0, _chai.expect)((0, _money.formatMoneyWithCode)(90734578, 'JPY')).to.be.equal('JPY 90 734 578.00');
    });
  });

  describe('formatMoneyDestructured', function () {
    it('destructures positive value', function () {
      (0, _chai.expect)((0, _money.formatMoneyDestructured)(123456, 'ZAR')).to.deep.equal({
        isoCode: 'ZAR',
        symbol: 'R',
        decimal: '.',
        units: ' 123 456',
        cents: '00',
        isNegative: false
      });
    });

    it('destructures negative value', function () {
      (0, _chai.expect)((0, _money.formatMoneyDestructured)(-123456, 'ZAR')).to.deep.equal({
        isoCode: 'ZAR',
        symbol: 'R',
        decimal: '.',
        units: '-123 456',
        cents: '00',
        isNegative: true
      });
    });

    it('destructures small negative value', function () {
      (0, _chai.expect)((0, _money.formatMoneyDestructured)(-0.75, 'ZAR')).to.deep.equal({
        isoCode: 'ZAR',
        symbol: 'R',
        decimal: '.',
        units: '-0',
        cents: '75',
        isNegative: true
      });
    });

    it('destructures with no cents', function () {
      (0, _chai.expect)((0, _money.formatMoneyDestructured)(123456, 'ZAR', { precision: 0 })).to.deep.equal({
        isoCode: 'ZAR',
        symbol: 'R',
        decimal: '.',
        units: ' 123 456',
        cents: undefined,
        isNegative: false
      });
    });

    it('destructures with overriden decimal separator', function () {
      (0, _chai.expect)((0, _money.formatMoneyDestructured)(123456, 'ZAR', { decimal: ',' })).to.deep.equal({
        isoCode: 'ZAR',
        symbol: 'R',
        decimal: ',',
        units: ' 123 456',
        cents: '00',
        isNegative: false
      });
    });
  });
});
