// @agct/react-common/dist/utils/formatting/numbers.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'accounting'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('accounting'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.accounting);
    global.numbers = mod.exports;
  }
})(this, function (exports, _accounting) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.formatDecimal = exports.formatNumber = exports.defaultNumberOptions = undefined;

  var _accounting2 = _interopRequireDefault(_accounting);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var defaultNumberOptions = exports.defaultNumberOptions = {
    precision: 0,
    thousand: ' ',
    decimal: '.'
  };

  var formatNumber = exports.formatNumber = function formatNumber(number) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
    return _accounting2.default.formatNumber(number, _extends({}, defaultNumberOptions, options));
  };

  var formatDecimal = exports.formatDecimal = function formatDecimal(number) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
    return formatNumber(number, _extends({
      precision: 2
    }, options));
  };
});
