// @agct/react-common/dist/components/Loading/Loading.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'classnames'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames);
    global.Loading = mod.exports;
  }
})(this, function (exports, _react, _classnames) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.LOADING_LARGE = exports.LOADING_SMALL = undefined;

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var LOADING_SMALL = exports.LOADING_SMALL = 'small';
  var LOADING_LARGE = exports.LOADING_LARGE = 'large';

  var Loading = function Loading(_ref) {
    var size = _ref.size;

    var baseClass = (0, _classnames2.default)('loading-spinner', {
      'loading-spinner--small': size === LOADING_SMALL,
      'loading-spinner--large': size === LOADING_LARGE
    });

    return _react2.default.createElement(
      'svg',
      { className: baseClass,
        xmlns: 'http://www.w3.org/2000/svg',
        viewBox: '0 0 66 66'
      },
      _react2.default.createElement('circle', { className: 'loading-spinner__path',
        cx: '33',
        cy: '33',
        r: '30',
        style: {
          fill: 'none',
          strokeWidth: 6,
          strokeLinecap: 'round'
        }
      })
    );
  };

  Loading.propTypes = {
    size: _react.PropTypes.string
  };

  exports.default = Loading;
});
