// @agct/react-common/dist/components/FormField/FormField.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'classnames', '@agct/reacceptance/src/qa', 'lodash'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'), require('@agct/reacceptance/src/qa'), require('lodash'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames, global.qa, global.lodash);
    global.FormField = mod.exports;
  }
})(this, function (exports, _react, _classnames, _qa, _lodash) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _lodash2 = _interopRequireDefault(_lodash);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var FormField = function FormField(_ref) {
    var field = _ref.field;
    var label = _ref.label;
    var prompt = _ref.prompt;
    var type = _ref.type;

    var showError = field.touched && field.error;

    var baseClass = (0, _classnames2.default)('input', {
      'is-error': showError });

    var fieldProps = _lodash2.default.omit(_extends({}, field), ['touched', 'error']);

    return _react2.default.createElement(
      'div',
      { className: 'input-wrapper' },
      _react2.default.createElement(
        'div',
        { className: baseClass },
        label && _react2.default.createElement(
          'label',
          null,
          label
        ),
        _react2.default.createElement('input', _extends({
          type: type,
          placeholder: prompt
        }, (0, _qa.attr)('input'), fieldProps)),
        _react2.default.createElement('div', { className: 'input__accent' })
      ),
      showError && _react2.default.createElement(
        'div',
        { className: 'input__error' },
        field.error
      )
    );
  };

  FormField.propTypes = {
    field: _react.PropTypes.object.isRequired,
    label: _react.PropTypes.string,
    prompt: _react.PropTypes.string,
    type: _react.PropTypes.string
  };

  FormField.defaultProps = {
    label: '',
    prompt: '',
    type: 'text'
  };

  exports.default = FormField;
});
