// @agct/react-common/src/components/AGOLink/AGOLink.jsx
import React, { PropTypes } from 'react';
import _ from 'lodash';
import { encode } from 'querystring';
import moment from 'moment';
import { attr as qa } from '@agct/reacceptance/src/qa';

const AGOLink = ({
  name,
  page,
  profileType,
  startDate,
  endDate,
  accountId,
  investorId,
  profileId,
  agoUrl,
  children,
  className,
}) => {
  const queries = encode(_.pickBy({
    redirect: page,
    ProfileType: profileType,
    StartDate: startDate && moment(startDate).format('YYYY-MM-DD'),
    EndDate: endDate && moment(endDate).format('YYYY-MM-DD'),
    AccountId: accountId,
    InvestorId: investorId,
    ProfileId: profileId,
  }));

  const url = `${agoUrl}/Oii.aspx?${queries}`;

  return (
    <a className={className} href={url} {...qa(name)}>
      { name }
      { children }
    </a>
    );
};

AGOLink.propTypes = {
  name: PropTypes.string,
  agoUrl: PropTypes.string.isRequired,
  children: PropTypes.any,
  page: PropTypes.string.isRequired,
  profileType: PropTypes.string.isRequired,

  startDate: PropTypes.any,
  endDate: PropTypes.any,
  accountId: PropTypes.string,
  investorId: PropTypes.string,
  profileId: PropTypes.string,
  className: PropTypes.string,
};

export default AGOLink;
