// @agct/react-common/dist/components/Button/Button.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', '@agct/reacceptance/src/qa'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('@agct/reacceptance/src/qa'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.qa);
    global.Button = mod.exports;
  }
})(this, function (exports, _react, _qa) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var Button = function Button(_ref) {
    var children = _ref.children;
    var _onClick = _ref.onClick;
    var _ref$type = _ref.type;
    var type = _ref$type === undefined ? 'button' : _ref$type;
    var _ref$isDisabled = _ref.isDisabled;
    var isDisabled = _ref$isDisabled === undefined ? false : _ref$isDisabled;
    var _ref$className = _ref.className;
    var className = _ref$className === undefined ? 'btn' : _ref$className;
    var _ref$qaName = _ref.qaName;
    var qaName = _ref$qaName === undefined ? 'button' : _ref$qaName;
    return _react2.default.createElement(
      'button',
      _extends({ className: className,
        disabled: isDisabled,
        onClick: function onClick() {
          return !isDisabled && _onClick && _onClick();
        },
        type: type
      }, (0, _qa.attr)(qaName)),
      children
    );
  };

  Button.propTypes = {
    onClick: _react.PropTypes.func,
    isDisabled: _react.PropTypes.bool,
    children: _react.PropTypes.any,
    type: _react.PropTypes.string,
    className: _react.PropTypes.string,
    qaName: _react.PropTypes.string
  };

  exports.default = Button;
});
