// @agct/react-common/dist/components/DropDown/DropDownUtils.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'lodash', './DropDownItem'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('lodash'), require('./DropDownItem'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.lodash, global.DropDownItem);
    global.DropDownUtils = mod.exports;
  }
})(this, function (exports, _react, _lodash, _DropDownItem) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.buildGroupOptions = exports.buildOptions = exports.buildBlankOption = undefined;

  var _react2 = _interopRequireDefault(_react);

  var _lodash2 = _interopRequireDefault(_lodash);

  var _DropDownItem2 = _interopRequireDefault(_DropDownItem);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var buildBlankOption = exports.buildBlankOption = function buildBlankOption(handleSelect) {
    var blankItem = { value: null, label: '--' };

    return _react2.default.createElement(_DropDownItem2.default, {
      item: blankItem,
      isSelected: false,
      isDisabled: false,
      onClick: handleSelect
    });
  };

  var buildOptions = exports.buildOptions = function buildOptions(items, selectedItem, handleSelect) {
    return items.map(function (item, key) {
      var isSelected = selectedItem && selectedItem.value === item.value;
      var isDisabled = _lodash2.default.get(item, 'disabled', false);

      return _react2.default.createElement(_DropDownItem2.default, {
        key: key,
        item: item,
        isSelected: isSelected,
        isDisabled: isDisabled,
        onClick: handleSelect
      });
    });
  };

  var buildGroupOptions = exports.buildGroupOptions = function buildGroupOptions(groups, selectedItem, handleSelect) {
    return groups.map(function (group, key) {
      var options = buildOptions(group.items, selectedItem, handleSelect);

      return _react2.default.createElement(
        'div',
        { key: key },
        _react2.default.createElement(
          'div',
          { className: 'dropdown__options__group' },
          group.label
        ),
        options
      );
    });
  };
});
