// @agct/react-common/dist/components/DropDown/DropDownItem.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'classnames', '@agct/reacceptance/src/qa'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'), require('@agct/reacceptance/src/qa'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames, global.qa);
    global.DropDownItem = mod.exports;
  }
})(this, function (exports, _react, _classnames, _qa) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var DropDownItem = function DropDownItem(_ref) {
    var item = _ref.item;
    var _onClick = _ref.onClick;
    var isDisabled = _ref.isDisabled;
    var isSelected = _ref.isSelected;

    var baseClass = (0, _classnames2.default)('dropdown__options__option', {
      'is-disabled': isDisabled,
      'is-selected': isSelected
    });

    return _react2.default.createElement(
      'div',
      _extends({
        className: baseClass,
        onClick: function onClick() {
          return _onClick(item);
        }
      }, (0, _qa.attr)('dropdown-option')),
      item.label
    );
  };

  DropDownItem.propTypes = {
    item: _react.PropTypes.object.isRequired,
    onClick: _react.PropTypes.func.isRequired,
    isDisabled: _react.PropTypes.bool,
    isSelected: _react.PropTypes.bool
  };

  exports.default = DropDownItem;
});
