// @agct/react-common/dist/components/DropDown/DropDownItem.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', './DropDownItem'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('./DropDownItem'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.DropDownItem);
    global.DropDownItemTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _DropDownItem) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _DropDownItem2 = _interopRequireDefault(_DropDownItem);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());

  describe('DropDown Item', function () {
    var dropdownItem = {
      key: 'item_1',
      label: 'item_label'
    };

    it('renders enabled unselected item correctly', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_DropDownItem2.default, {
        key: dropdownItem.key,
        item: dropdownItem,
        isSelected: false,
        isDisabled: false,
        onClick: function onClick() {}
      }));
      (0, _chai.expect)(wrapper).to.have.className('dropdown__options__option');
      (0, _chai.expect)(wrapper).to.not.have.className('is-selected');
      (0, _chai.expect)(wrapper).to.not.have.className('is-disabled');
      (0, _chai.expect)(wrapper).to.have.text('item_label');
    });

    it('renders disabled unselected item correctly', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_DropDownItem2.default, {
        key: dropdownItem.key,
        item: dropdownItem,
        isSelected: false,
        isDisabled: true,
        onClick: function onClick() {}
      }));
      (0, _chai.expect)(wrapper).to.have.className('dropdown__options__option');
      (0, _chai.expect)(wrapper).to.not.have.className('is-selected');
      (0, _chai.expect)(wrapper).to.have.className('is-disabled');
      (0, _chai.expect)(wrapper).to.have.text('item_label');
    });

    it('renders enabled selected item correctly', function () {
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_DropDownItem2.default, {
        key: dropdownItem.key,
        item: dropdownItem,
        isSelected: true,
        isDisabled: false,
        onClick: function onClick() {}
      }));
      (0, _chai.expect)(wrapper).to.have.className('dropdown__options__option');
      (0, _chai.expect)(wrapper).to.have.className('is-selected');
      (0, _chai.expect)(wrapper).to.not.have.className('is-disabled');
      (0, _chai.expect)(wrapper).to.have.text('item_label');
    });
  });
});
