// @agct/react-common/dist/components/DropDown/DropDown.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', './DropDown'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('./DropDown'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.DropDown);
    global.DropDownTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _DropDown) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _DropDown2 = _interopRequireDefault(_DropDown);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());

  describe('Dropdown Component', function () {
    it('should render when no options are presented', function () {
      var actions = {
        onClick: function onClick() {},
        onFocus: function onFocus() {},
        onChange: function onChange() {}
      };
      var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_DropDown2.default, { onChange: actions.onChange, onFocus: actions.onFocus }));
      var dropdown = wrapper.find('.dropdown');
      (0, _chai.expect)(dropdown.childAt(0)).to.have.className('btn');
      (0, _chai.expect)(dropdown.childAt(0)).to.have.className('dropdown__toggle');
      (0, _chai.expect)(dropdown.childAt(1)).to.have.className('dropdown__options');
      (0, _chai.expect)(wrapper.find('.dropdown__title')).to.have.text('Please select a value');
    });
  });

  describe('Dropdown Actions', function () {
    var actions = {
      onClick: function onClick() {},
      onFocus: function onFocus() {},
      onChange: function onChange() {}
    };
    var dropdownItems = [{ value: 'GBP', label: 'Pounds', disabled: true }, { value: 'USD', label: 'Dollars' }];

    var wrapper = (0, _enzyme.mount)(_react2.default.createElement(_DropDown2.default, { handleToggle: actions.onClick,
      onChange: actions.onChange, onFocus: actions.onFocus,
      items: dropdownItems
    }));

    it('should toggle open/close when clicked', function () {
      var dropdownBtn = wrapper.find('.btn');
      (0, _chai.expect)(wrapper.find('.dropdown__options')).to.not.have.className('is-open');
      dropdownBtn.simulate('click');
      (0, _chai.expect)(wrapper.find('.dropdown__options')).to.have.className('is-open');
      dropdownBtn.simulate('click');
      (0, _chai.expect)(wrapper.find('.dropdown__options')).to.not.have.className('is-open');
    });

    it('should not toggle open when clicked if dropdown disabled', function () {
      var disabledDropDown = (0, _enzyme.mount)(_react2.default.createElement(_DropDown2.default, { handleToggle: actions.onClick,
        onChange: actions.onChange, onFocus: actions.onFocus,
        items: dropdownItems, isDisabled: true
      }));
      var dropdownBtn = disabledDropDown.find('.btn');
      (0, _chai.expect)(wrapper.find('.dropdown__options')).to.not.have.className('is-open');
      dropdownBtn.simulate('click');
      (0, _chai.expect)(wrapper.find('.dropdown__options')).to.not.have.className('is-open');
    });

    it('should list options passed through to it', function () {
      var listOptions = wrapper.find('.dropdown__options');
      (0, _chai.expect)(listOptions.children()).to.have.length(dropdownItems.length + 1);
      (0, _chai.expect)(listOptions.childAt(0)).to.have.text('--');
      (0, _chai.expect)(listOptions.childAt(1)).to.have.text('Pounds');
      (0, _chai.expect)(listOptions.childAt(1)).to.have.className('dropdown__options__option');
      (0, _chai.expect)(listOptions.childAt(1)).to.have.className('is-disabled');
      (0, _chai.expect)(listOptions.childAt(2)).to.have.text('Dollars');
      (0, _chai.expect)(listOptions.childAt(2)).to.have.className('dropdown__options__option');
      (0, _chai.expect)(listOptions.childAt(2)).to.not.have.className('is-disabled');
    });
  });
});
