// @agct/react-common/src/components/TabsMenuSlider/TabsMenuSliderArrows.jsx
import React, { PropTypes } from 'react';
import classnames from 'classnames';

const TabsMenuSliderArrows = ({ tabsLength, activePage, tabsPerPage, displayArrows, prevPage, nextPage }) => {
  const prevArrowClass = classnames('tabs-slider__nav tabs-slider__nav--prev', {
    'is-disabled': activePage === 0,
  });
  const nextArrowClass = classnames('tabs-slider__nav tabs-slider__nav--next', {
    'is-disabled': activePage === tabsLength - tabsPerPage,
  });

  if (!displayArrows) return false;

  return (
    <div className="tabs-slider__arrows">
      <div className={prevArrowClass} onClick={prevPage}>
        <i className="agi agi agi-double-chevron-left"></i>
      </div>
      <div className={nextArrowClass} onClick={nextPage}>
        <i className="agi agi agi-double-chevron-right"></i>
      </div>
    </div>
  );
};

TabsMenuSliderArrows.propTypes = {
  tabsLength: PropTypes.number.isRequired,
  activePage: PropTypes.number.isRequired,
  tabsPerPage: PropTypes.number.isRequired,
  displayArrows: PropTypes.bool.isRequired,
  prevPage: PropTypes.func.isRequired,
  nextPage: PropTypes.func.isRequired,
};

export default TabsMenuSliderArrows;
