// @agct/react-common/dist/components/TabsMenuSlider/TabsMenuSlider.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', 'sinon-chai', './Variants', './TabsMenuSlider'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('sinon-chai'), require('./Variants'), require('./TabsMenuSlider'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.sinonChai, global.Variants, global.TabsMenuSlider);
    global.TabsMenuSliderTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _sinonChai, _Variants, _TabsMenuSlider) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _sinonChai2 = _interopRequireDefault(_sinonChai);

  var _TabsMenuSlider2 = _interopRequireDefault(_TabsMenuSlider);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());
  _chai2.default.use(_sinonChai2.default);

  var props = {
    tabs: [{
      title: function title() {
        return 'tab 1';
      },
      content: function content() {
        return 'tab 1 content';
      }
    }, {
      title: function title() {
        return 'tab 2';
      },
      content: function content() {
        return 'tab 2 content';
      }
    }, {
      title: function title() {
        return 'tab 3';
      },
      content: function content() {
        return 'tab 3 content';
      }
    }, {
      title: function title() {
        return 'tab 4';
      },
      content: function content() {
        return 'tab 4 content';
      }
    }],
    activeTabIndex: 0,
    activeTabContent: 'tab 1 content',
    onChange: function onChange() {}
  };
  var wrapper = (0, _enzyme.shallow)(_react2.default.createElement(_TabsMenuSlider2.default, {
    tabs: props.tabs.map(function (tab) {
      return tab.header;
    }),
    activeTabIndex: props.activeTabIndex,
    activeTabContent: props.activeTabContent,
    onChange: props.onChange
  }));

  describe('TabsMenuSlider component', function () {
    it('renders tabs', function () {
      (0, _chai.expect)(wrapper.find('.tabs-slider__tab')).to.have.length(props.tabs.length);
    });

    it('renders active tab', function () {
      (0, _chai.expect)(wrapper.find('.tabs-slider__tab.is-active')).to.have.length(1);
      (0, _chai.expect)(wrapper.find('.tabs-slider__tab').at(0)).to.have.className('is-active');
      (0, _chai.expect)(wrapper.find('.tabs-slider__detail').text()).to.equal(props.tabs[0].content());
    });

    it('renders variants', function () {
      var wrapperArrow = (0, _enzyme.shallow)(_react2.default.createElement(_TabsMenuSlider2.default, {
        tabsVariant: _Variants.TABS_VARIANT_ARROW,
        tabs: props.tabs.map(function (tab) {
          return tab.header;
        }),
        activeTabIndex: props.activeTabIndex,
        activeTabContent: props.activeTabContent,
        onChange: props.onChange
      }));
      var wrapperUnderline = (0, _enzyme.shallow)(_react2.default.createElement(_TabsMenuSlider2.default, {
        tabsVariant: _Variants.TABS_VARIANT_UNDERLINE,
        tabs: props.tabs.map(function (tab) {
          return tab.header;
        }),
        activeTabIndex: props.activeTabIndex,
        activeTabContent: props.activeTabContent,
        onChange: props.onChange
      }));
      (0, _chai.expect)(wrapperArrow.find('.tabs-slider__tabs--down-arrow')).to.exist;
      (0, _chai.expect)(wrapperUnderline.find('.tabs-slider__tabs--underline')).to.exist;
    });

    describe('next and prev', function () {
      var activePage = void 0;
      var sandbox = sinon.sandbox.create();
      var spyNextPage = sandbox.spy(_TabsMenuSlider2.default.prototype, 'nextPage');
      var spyPrevPage = sandbox.spy(_TabsMenuSlider2.default.prototype, 'prevPage');
      var wrapperMount = (0, _enzyme.mount)(_react2.default.createElement(_TabsMenuSlider2.default, {
        tabs: props.tabs.map(function (tab) {
          return tab.header;
        }),
        activeTabIndex: props.activeTabIndex,
        activeTabContent: props.activeTabContent,
        onChange: props.onChange
      }));

      wrapperMount.find('.tabs-slider__nav--next').simulate('click');
      (0, _chai.expect)(spyNextPage).to.have.been.calledOnce;
      activePage = wrapperMount.state('activePage');
      (0, _chai.expect)(activePage).to.equal(1);

      wrapperMount.find('.tabs-slider__nav--next').simulate('click');
      activePage = wrapperMount.state('activePage');
      (0, _chai.expect)(activePage).to.equal(2);

      wrapperMount.find('.tabs-slider__nav--prev').simulate('click');
      (0, _chai.expect)(spyPrevPage).to.have.been.calledOnce;
      activePage = wrapperMount.state('activePage');
      (0, _chai.expect)(activePage).to.equal(1);
      sandbox.restore();
    });

    describe('actions - onChange', function () {
      var sandbox = sinon.sandbox.create();
      var spyOnChange = sandbox.spy(props, 'onChange');
      var wrapperMount = (0, _enzyme.mount)(_react2.default.createElement(_TabsMenuSlider2.default, {
        tabs: props.tabs.map(function (tab) {
          return tab.header;
        }),
        activeTabIndex: props.activeTabIndex,
        activeTabContent: props.activeTabContent,
        onChange: props.onChange
      }));

      wrapperMount.find('.tabs-slider__tab').at(2).simulate('click');
      (0, _chai.expect)(spyOnChange).to.have.been.calledOnce;
      (0, _chai.expect)(spyOnChange).to.have.been.calledWith(2);
      sandbox.restore();
    });
  });
});
