// @agct/react-common/dist/components/TabsMenuSlider/TabsMenuSliderArrows.test.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['react', 'chai', 'enzyme', 'chai-enzyme', './TabsMenuSliderArrows'], factory);
  } else if (typeof exports !== "undefined") {
    factory(require('react'), require('chai'), require('enzyme'), require('chai-enzyme'), require('./TabsMenuSliderArrows'));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.react, global.chai, global.enzyme, global.chaiEnzyme, global.TabsMenuSliderArrows);
    global.TabsMenuSliderArrowsTest = mod.exports;
  }
})(this, function (_react, _chai, _enzyme, _chaiEnzyme, _TabsMenuSliderArrows) {
  'use strict';

  var _react2 = _interopRequireDefault(_react);

  var _chai2 = _interopRequireDefault(_chai);

  var _chaiEnzyme2 = _interopRequireDefault(_chaiEnzyme);

  var _TabsMenuSliderArrows2 = _interopRequireDefault(_TabsMenuSliderArrows);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  _chai2.default.use((0, _chaiEnzyme2.default)());

  var props = {
    tabsLength: 10,
    activePage: 0,
    tabsPerPage: 5,
    displayArrows: true,
    prevPage: function prevPage() {},
    nextPage: function nextPage() {}
  };

  var wrapperDisplayArrows = (0, _enzyme.shallow)(_react2.default.createElement(_TabsMenuSliderArrows2.default, {
    tabsLength: props.tabsLength,
    activePage: props.activePage,
    tabsPerPage: props.tabsPerPage,
    displayArrows: props.tabsPerPage,
    prevPage: props.prevPage,
    nextPage: props.nextPage
  }));

  var wrapperNoArrows = (0, _enzyme.shallow)(_react2.default.createElement(_TabsMenuSliderArrows2.default, {
    tabsLength: props.tabsLength,
    activePage: props.activePage,
    tabsPerPage: props.tabsPerPage,
    displayArrows: !props.tabsPerPage,
    prevPage: props.prevPage,
    nextPage: props.nextPage
  }));

  describe('TabsMenuSliderArrows component', function () {
    it('renders arrows if displayArrows is true', function () {
      (0, _chai.expect)(wrapperDisplayArrows.find('.tabs-slider__arrows')).to.exist;
    });

    it('does not render arrows if displayArrows is false', function () {
      (0, _chai.expect)(wrapperNoArrows.find('.tabs-slider__arrows')).to.not.exist;
    });
  });
});
