// @agct/react-common/dist/components/TabsMenuSlider/TabsMenuSliderArrows.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'classnames'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames);
    global.TabsMenuSliderArrows = mod.exports;
  }
})(this, function (exports, _react, _classnames) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var TabsMenuSliderArrows = function TabsMenuSliderArrows(_ref) {
    var tabsLength = _ref.tabsLength;
    var activePage = _ref.activePage;
    var tabsPerPage = _ref.tabsPerPage;
    var displayArrows = _ref.displayArrows;
    var prevPage = _ref.prevPage;
    var nextPage = _ref.nextPage;

    var prevArrowClass = (0, _classnames2.default)('tabs-slider__nav tabs-slider__nav--prev', {
      'is-disabled': activePage === 0
    });
    var nextArrowClass = (0, _classnames2.default)('tabs-slider__nav tabs-slider__nav--next', {
      'is-disabled': activePage === tabsLength - tabsPerPage
    });

    if (!displayArrows) return false;

    return _react2.default.createElement(
      'div',
      { className: 'tabs-slider__arrows' },
      _react2.default.createElement(
        'div',
        { className: prevArrowClass, onClick: prevPage },
        _react2.default.createElement('i', { className: 'agi agi agi-double-chevron-left' })
      ),
      _react2.default.createElement(
        'div',
        { className: nextArrowClass, onClick: nextPage },
        _react2.default.createElement('i', { className: 'agi agi agi-double-chevron-right' })
      )
    );
  };

  TabsMenuSliderArrows.propTypes = {
    tabsLength: _react.PropTypes.number.isRequired,
    activePage: _react.PropTypes.number.isRequired,
    tabsPerPage: _react.PropTypes.number.isRequired,
    displayArrows: _react.PropTypes.bool.isRequired,
    prevPage: _react.PropTypes.func.isRequired,
    nextPage: _react.PropTypes.func.isRequired
  };

  exports.default = TabsMenuSliderArrows;
});
