// @agct/react-common/src/components/FormSwitch/FormSwitch.jsx
import React, { PropTypes } from 'react';
import _ from 'lodash';

class FormSwitch extends React.Component {
  constructor(props) {
    super(props);
    this.state = { ...props };
  }

  componentWillMount() {
    const id = _.uniqueId('switch');
    this.setState({ id });
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ ...nextProps });
  }

  handleChange(event) {
    this.setState({
      isChecked: !this.state.isChecked,
    });
    this.props.onChange(event);
  }

  render() {
    return (
      <div className="switch-wrap">
        <input type="checkbox"
          id={this.state.id}
          value={this.state.id}
          checked={this.state.isChecked}
          disabled={this.state.isDisabled}
          onChange={(e) => !this.state.isDisabled && this.handleChange(e)}
        ></input>
        <label className="switch" htmlFor={this.state.id}>
          <span className="switch__toggle"></span>
          <span className="switch__label">{this.state.label}</span>
        </label>
      </div>
    );
  }
}

FormSwitch.propTypes = {
  label: PropTypes.string.isRequired,
  isDisabled: PropTypes.bool,
  isChecked: PropTypes.bool,
  onChange: PropTypes.func,
};

FormSwitch.defaultProps = {
  isChecked: false,
  isDisabled: false,
  onChange: () => {},
};

export default FormSwitch;
