// @agct/react-common/dist/components/AGOLink/AGOLink.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'lodash', 'querystring', 'moment', '@agct/reacceptance/src/qa'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('lodash'), require('querystring'), require('moment'), require('@agct/reacceptance/src/qa'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.lodash, global.querystring, global.moment, global.qa);
    global.AGOLink = mod.exports;
  }
})(this, function (exports, _react, _lodash, _querystring, _moment, _qa) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _lodash2 = _interopRequireDefault(_lodash);

  var _moment2 = _interopRequireDefault(_moment);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  var AGOLink = function AGOLink(_ref) {
    var name = _ref.name;
    var page = _ref.page;
    var profileType = _ref.profileType;
    var startDate = _ref.startDate;
    var endDate = _ref.endDate;
    var accountId = _ref.accountId;
    var investorId = _ref.investorId;
    var profileId = _ref.profileId;
    var agoUrl = _ref.agoUrl;
    var children = _ref.children;
    var className = _ref.className;

    var queries = (0, _querystring.encode)(_lodash2.default.pickBy({
      redirect: page,
      ProfileType: profileType,
      StartDate: startDate && (0, _moment2.default)(startDate).format('YYYY-MM-DD'),
      EndDate: endDate && (0, _moment2.default)(endDate).format('YYYY-MM-DD'),
      AccountId: accountId,
      InvestorId: investorId,
      ProfileId: profileId
    }));

    var url = agoUrl + '/Oii.aspx?' + queries;

    return _react2.default.createElement(
      'a',
      _extends({ className: className, href: url }, (0, _qa.attr)(name)),
      name,
      children
    );
  };

  AGOLink.propTypes = {
    name: _react.PropTypes.string,
    agoUrl: _react.PropTypes.string.isRequired,
    children: _react.PropTypes.any,
    page: _react.PropTypes.string.isRequired,
    profileType: _react.PropTypes.string.isRequired,

    startDate: _react.PropTypes.any,
    endDate: _react.PropTypes.any,
    accountId: _react.PropTypes.string,
    investorId: _react.PropTypes.string,
    profileId: _react.PropTypes.string,
    className: _react.PropTypes.string
  };

  exports.default = AGOLink;
});
